provisioning/miniprov/hare_mp/store.py

```diff
@@ -25,8 +25,11 @@
 
 
 class ValueProvider:
-    def get(self, key: str) -> Any:
-        return self._raw_get(key)
+    def get(self, key: str, allow_null: bool = False) -> Any:
+        ret = self._raw_get(key)
+        if ret is None and not allow_null:
```

Comment:
I think we need to set expectations clearly here, `allow_null` is for us, hare. We know that this KV is not provided by conf-store and thus Hare is setting it to `null` by default. But if the KV is present in conf-store then it cannot be null I believe.  This routine is used to fetch KV from conf-store, that means Hare knows that the given key is present in conf-store. It cannot be `null` if the value is `None` we must raise an exception I think. cc @knekrasov.

Reply:
@mssawant This piece of code is actually mine 😁 

The fact is that ConfStore's get() method returns None if the JSON file doesn't have that value. So either our code is incorrect (i.e. the key name is misspelled), or ConfStore is not full (we saw that as well) or our code doesn't correspond ConfStore. That's why I insisted to throw an exception if ConfStore returns None by default.

But there is at least one situation when ConfStore can return None - iface_type. It can be omitted. Then the caller will need to put `allow_null=True`. But that's a kind of special case and the caller code is quite explicit about that.

Reply:
Right, we need to raise an exception for `None` returned from conf-store.

But if `iface_type` key is present in conf-store then why would conf-store not have a value to assigned to it? I think we need to take this up with conf-store team to set a default value instead of returning `None` for it if the key is present. Problem here is that if we allow the keys to `None` we need to understand the circumstances, the contract needs to be more clear, no?

Reply:
Not sure. According to Dhall types, iface_type is optional.  So it can be None. And thus it can be omitted in CDF file, CFGEN will take care of that and interpret it according to its own logic. I don't see a reason why we should make a decision on default values at the level of cdf generator.

Reply:
> But if iface_type key is present in conf-store then why would conf-store not have a value assigned to it? 

ConfStore can miss that key. That is the reason why ConfStore API can return None.